### PR TITLE
Add friend functionality

### PR DESCRIPTION
**Pull request by sl7v3**
Functionality has been added to add friends, accept request, view all friends, active friends, pending requests. 
A few issues like displaying color icons for online status, scrolling of friends list page has not been implemented and will be moved to another issue.
Review and let me know of any issues that need to be addressed.

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -13,5 +13,8 @@
 //= require jquery
 //= require bootstrap-sprockets
 //= require jquery_ujs
+//This turbolinks gem causes submissions to be made twice. If anyone runs into an error where their submit button tries to
+//perform an action twice even though you clicked it once, you may remove turbolinks (I have not used it anyway) or you may
+// add the data-no-turbolink tag to your html object.
```
Comment (danfstucky) [inline app/assets/javascripts/application.js]: Did you add this comment and find this issue?  If you found a defect, log an issue for an it

File: app/assets/stylesheets/application.css.scss [hunk 2]
```diff
@@ -220,25 +220,19 @@ td {
...
       width: 100%;
     }
-    .btn-default{
+
+    .btn-default {
 	    color: white;
 	    background: black;
 	  }
-	  .champ-img{
+    .champ-img{
```
Comment (danfstucky) [inline app/assets/stylesheets/application.css.scss]: Try to keep your spacing consistent everywhere.  You added a space between the btn class and bracket above, but didn't do it here and other areas.  There should be a space for all css selectors

File: app/assets/stylesheets/friendships.scss [hunk 3]
```diff
@@ -0,0 +1,44 @@
+.summoner-request-page, .summoner-friends{
+	position: fixed;
+	width: 100%;
+	min-height: 1080px;
+	/*height: 100%;*/
```
Comment (danfstucky) [inline app/assets/stylesheets/friendships.scss]: remove comment

File: app/assets/stylesheets/friendships.scss [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	background-image: image-url("login_bg.png");
+	background-repeat: no-repeat;
+	background-size: cover;
+	padding-top: 20px;
+	.request-table, .friends-list-table{
+    border-collapse: collapse;
+    margin-top:60px;
+	  border: 1px solid black;
+	  cellpadding: 0;
```
Comment (danfstucky) [inline app/assets/stylesheets/friendships.scss]: I suspect you can remove this cellpadding and it will have no effect.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -5,6 +5,11 @@ class ApplicationController < ActionController::Base
   include SessionsHelper
 
 #Including some general functions here so they can be reached by several controllers
-
+	def require_user
+	  if !logged_in?
+	    flash[:danger] = "You must be logged in to perform that action"
+	    redirect_to root_path
+	  end
+	end
```
Comment (danfstucky) [inline app/controllers/application_controller.rb]: 1) Include a brief comment about what the function does.
2) There is no reason to have this function here.  It is related to a session and is using another session function so it belongs inside the SessionsHelper.  Since the helper is included in application_controller all those functions are available to every controller anyway

File: app/controllers/friendships_controller.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
+class FriendshipsController < ApplicationController
+  include LolConnections
+  before_action :search_summoner_for_request_action, only: [:edit]
+  before_action :require_user
+  before_action :require_same_user, only: [:edit, :decide]
+  
+ 
+	def new
```
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: indentation is off here

File: app/controllers/friendships_controller.rb [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    @user = current_user
```
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: You really need to look at how you are using instance variables in this controller.  Instance variables start with an @ sign.  You should keep the scope of a variable as tight as possible so if you don't need to use a variable outside of a particular function, then it should be local.  A lot of the variables here are only used in the function they are declared in so there is no reason to make them belong to the entire instance.  I'm not even sure how some of this is working correctly.  You redefine @user 4 times and you are even using the same definition for some of them.  
For example, if you wanted to get the current user you could do `@user ||= current_user`.  This means, it will only execute the current_user function if @user is not already defined.  You don't want to redo any calls that you don't have to.  Many of these (including `current_user`) require db queries and are not cheap to execute.  Reevaluate how you are using ALL your variables in this controller and make changes as you deem necessary.

File: app/controllers/friendships_controller.rb [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    if @friends_list_code == 'all'
+      @friend_count = @user.accepted_friendships.count
+      @friendships = @user.friendships.accepted.paginate(page: params[:page], per_page: 5)
+    elsif @friends_list_code == 'active'
+      @friendships = @user.get_online_friends.paginate(page: params[:page], per_page: 5)
+    else 
+      @pending_friendships_count = @user.pending_friendships.count
+      @friendships = @user.friendships.pending.paginate(page: params[:page], per_page: 5)
+    end
```
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: Have a final statement here to handle the pagination.  `@friendships.paginate(blah blah)` instead of writing the paginate method 3 times

File: app/controllers/search_controller.rb [hunk 9]
```diff
@@ -3,7 +3,7 @@ class SearchController < ApplicationController
   before_action :search_summoner, only: [:search]
   
   def search
-    @name = params[:summoner]
+    @name = params[:summoner].downcase
     if @name.to_s.length<=1
```
Comment (danfstucky) [inline app/controllers/search_controller.rb]: Not sure, but is the `.to_s` needed here?  I don't think you could call downcase if it wasn't already a string

File: db/migrate/20160726034132_add_last_logged_in_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddLastLoggedInToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :last_logged_in, :datetime
```
Comment (danfstucky) [inline db/migrate/20160726034132_add_last_logged_in_to_users.rb]: Why do we need to know when a user last logged in?  Are you using this somewhere else?

File: app/models/friendship.rb [hunk 11]
```diff
@@ -0,0 +1,67 @@
+class Friendship < ActiveRecord::Base
+  extend TokenContext
+  include Tokenizer
+	enum friendship_status: [ :denied, :pending, :accepted]
+  attr_accessor :friendship_request_token
+	cattr_accessor :daily_request_limit
```
Comment (danfstucky) [inline app/models/friendship.rb]: Typo here.  I am skeptical whether you even need an accessor for daily_request_limit since it is definitely not working here and you declare it as a class variable below

File: app/models/friendship.rb [hunk 12]
```diff
@@ -0,0 +1,67 @@
+class Friendship < ActiveRecord::Base
+  extend TokenContext
+  include Tokenizer
+	enum friendship_status: [ :denied, :pending, :accepted]
```
Comment (danfstucky) [inline app/models/friendship.rb]: It's awesome that you're using an enum here.  I feel like this may also have been useful in your friendships controller.  Not sure though, and I'm too lazy to look back and confirm that...

File: app/models/friendship.rb [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  attr_accessor :friendship_request_token
+	cattr_accessor :daily_request_limit
+  @@daily_request_limit = 12
+  belongs_to :user, :foreign_key => "user_id", :class_name => "User"
```
Comment (danfstucky) [inline app/models/friendship.rb]: Use the hash syntax here, not the rocket.  `foreign_key: 'user_id', etc.
Also, you should use single quotes when possible.  In ruby double quotes are more expensive because the interpreter has to check if any string interpolation is occurring. 
Lastly, your indent is off for the relation below this.

File: app/models/friendship.rb [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	end
+
+  def create_friendship_request_digest
+    self.friendship_request_token = Friendship.new_token
+    self.friendship_request_digest = Friendship.digest(friendship_request_token)
+  end
+
+  def accept_friend_request
+    update_attribute(:friendship_status, 2)
+    update_attribute(:request_responded_at, Time.zone.now)
```
Comment (danfstucky) [inline app/models/friendship.rb]: I could be wrong, but I think you can do this in one update with something like
`update(friendship_status: 2, request_responded_at: Time.zone.now)`
Same for the below update

File: app/models/user.rb [hunk 15]
```diff
@@ -75,18 +65,29 @@ def self.find_summoner_by_name(summoner_name)
...
+  end
+
 	private
 
 	#Converts email to all lower-case
 	def downcase_email
 		self.email = email.downcase
 	end
 
+	#Converts email to all lower-case
```
Comment (danfstucky) [inline app/models/user.rb]: Your documentation is off for this function.  Also, you could probably combine the email and name downcase into a single function since they are only ever used together

File: app/views/friendships/_friends_list.html.erb [hunk 16]
```diff
@@ -0,0 +1,39 @@
+<% obj.each do |friendship| %>
```
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: `obj` is about as nondescriptive of a parameter name as you can get. That could literally be anything

File: app/views/friendships/_friends_list.html.erb [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+<tbody class = "friend-row">
+	<tr>
+		<td ><strong>Summoner Name: <%= friendship.friend.name.capitalize %></strong></td>
+		<% if @friends_list_code == 'pending' %> 
+		  <td ><strong> <%if friendship.initiator == false %>
+		                  Received: 
+		                <% else %>
+		                  Sent: 
+		                <% end %> 
+		    <%= "#{time_ago_in_words(friendship.created_at)}".capitalize %> ago</strong></td>
```
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: You don't need string interpolation for all these method calls.  That just slows the interpreter down.
You only need interpolation if you want to combine a string with a method call.  For example:
`"Hello, my name is #{user.name_var}"`.  If you just want the value of my name, you would just call `user.name` with no interpolation.  Check all of the views you made in this pull request and remove any unnecessary string interpolation.

File: app/views/friendships/_friends_list.html.erb [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+    </td>
+	</tr>
+	<tr>
+		<td ><strong>Friendship Status: <%= friendship.friendship_status.capitalize %></strong></td>
+	  <td ><strong>Member Since: <%= "#{time_ago_in_words(friendship.friend.created_at)}".capitalize %> ago</strong></td>
+	</tr>
+	<tr data-no-turbolink>
+	  <% if @friends_list_code == 'pending' %> 
+	    <%if friendship.initiator == false %>
+		    <td ><%= link_to  "Accept Request", decide_friendships_path(friend_email: current_user.email, requester_email: User.find(friendship.friend.id).email, request_token: 'from-friends-list', response_code: 'accept'), :class =>"btn btn-success" %> </td>
```
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: You really shouldn't be using hash rocket syntax anywhere.  That is old syntax.  This can be made into `class: "btn btn-success"`

File: app/views/friendships/_friends_list.html.erb [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+		    <td>	<%= link_to  "Deny Request", decide_friendships_path(friend_email: current_user.email, requester_email: User.find(friendship.friend.id).email, request_token: 'from-friends-list', response_code: 'deny'), :class =>"btn btn-danger" %>
+		    </td>
+		  <% else %>
+		    <td width="100%"><%= link_to "View Summoner Profile", profile_path(friendship.friend.id), :class =>"btn btn-success" %></td>
```
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: Why is there css in the view?!

File: app/views/friendships/edit.html.erb [hunk 20]
```diff
@@ -0,0 +1,49 @@
+<div class = "summoner-request-page" id="summoner-request">
```
Comment (danfstucky) [inline app/views/friendships/edit.html.erb]: Your spacing is very inconsistent when using these css classes.  In general there should not be spaces between the class keyword and the class name ever.  `class="summoner-request-page" id="summoner-request"`.  Look throughout all your views in this pull request and make sure your spacing is consistent.

File: app/views/friendships/index.html.erb [hunk 21]
```diff
@@ -0,0 +1,29 @@
+<div class = "summoner-friends" id="summoner-friends-list">
+  <div class = "col-md-5 col-md-offset-3">
+    <div class="table-responsive">
+    	<table class="table friends-list-table" cellspacing="0" style="border-spacing: 0;">
```
Comment (danfstucky) [inline app/views/friendships/index.html.erb]: Get rid of this styling in your html.  Also, never use cellspacing="0".  `border-collapse: collapse;` does the exact same thing

File: test/fixtures/friendships.yml [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+
+one:
+  id: 1
+  user_id: 1
+  friend_id: 1
+
+two:
+  id: 1
+  user_id: 1
+  friend_id: 1
```
Comment (danfstucky) [inline test/fixtures/friendships.yml]: Why did you add information to the fixture if you don't actually have a test that uses it?

File: app/assets/stylesheets/application.css.scss (hunk 2 above)
Comment (sl7v3) [inline app/assets/stylesheets/application.css.scss]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/f7bba8ef8acf1ed4f5dbf551bca18ee83220b20d

File: app/assets/stylesheets/friendships.scss (hunk 3 above)
Comment (sl7v3) [inline app/assets/stylesheets/friendships.scss]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/f7bba8ef8acf1ed4f5dbf551bca18ee83220b20d

File: app/assets/stylesheets/friendships.scss (hunk 4 above)
Comment (sl7v3) [inline app/assets/stylesheets/friendships.scss]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/f7bba8ef8acf1ed4f5dbf551bca18ee83220b20d

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (sl7v3) [inline app/controllers/application_controller.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/pull/42/commits/0ab6d0fc60b4a42eb2d34f1ad0788ec265af22e2

File: app/controllers/friendships_controller.rb (hunk 6 above)
Comment (sl7v3) [inline app/controllers/friendships_controller.rb]: This indentation looked good in my editor....I double-checked when making this commit - https://github.com/danfstucky/LeagueSpec/commit/60b313a63b15ac10c4effcb6c16158219ace0d88

File: app/controllers/friendships_controller.rb (hunk 7 above)
Comment (sl7v3) [inline app/controllers/friendships_controller.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/60b313a63b15ac10c4effcb6c16158219ace0d88

File: app/controllers/friendships_controller.rb (hunk 8 above)
Comment (sl7v3) [inline app/controllers/friendships_controller.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/fff87f252fe8b8258a789becf0a60806ac0eba32

File: app/controllers/search_controller.rb (hunk 9 above)
Comment (sl7v3) [inline app/controllers/search_controller.rb]: Fix applied to this and a few other cases where downcase was used: https://github.com/danfstucky/LeagueSpec/commit/7b7cce83e3b3bc9cda92fa24ecc5fe1b5d87fc7b

File: db/migrate/20160726034132_add_last_logged_in_to_users.rb (hunk 10 above)
Comment (sl7v3) [inline db/migrate/20160726034132_add_last_logged_in_to_users.rb]: I am not using this feature right now, but I plan to sort users by their last activity on LeagueSpec and it will be used then. It will also be used to apply the Yellow Activity Icon if they have been recently active but not currently online

File: app/models/friendship.rb (hunk 11 above)
Comment (sl7v3) [inline app/models/friendship.rb]: I do not see any typo here. My indentation is also fine in my editor, I don't know why it is looking different here.
The cattr_accessor is meant to do what it does - provides access to a class variable that prevents users from adding more than 12 users. It is used in user.rb line 68 and friendship.rb line 21

File: app/models/friendship.rb (hunk 12 above)
Comment (sl7v3) [inline app/models/friendship.rb]: It might have been useful but it doesn't make any difference since I'll still have to check for the same number of conditions.

File: app/models/friendship.rb (hunk 13 above)
Comment (sl7v3) [inline app/models/friendship.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/a8a6100dee7bf3e12c2083b5536c45c3a4329790

Comment (sl7v3) [inline app/models/friendship.rb]: Indentation looks good on my end. Fix applied here: https://github.com/danfstucky/LeagueSpec/commit/2bc01b6b730ab71c6b9aca1505cd0a68aa688697

File: app/models/friendship.rb (hunk 14 above)
Comment (sl7v3) [inline app/models/friendship.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/29a070cce4385e1711cd9d725659c77924628318

File: app/models/user.rb (hunk 15 above)
Comment (sl7v3) [inline app/models/user.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/54e7caa097ea5e3b42b4b130f679e54459cc1db8

File: app/views/friendships/_friends_list.html.erb (hunk 17 above)
Comment (sl7v3) [inline app/views/friendships/_friends_list.html.erb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/1170d0f3e9d7b58c66d4c54c63c493552b794230

File: app/views/friendships/_friends_list.html.erb (hunk 16 above)
Comment (sl7v3) [inline app/views/friendships/_friends_list.html.erb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/786e4c56ed3380482ca2071a1ca4c73969e5afb2

File: app/views/friendships/_friends_list.html.erb (hunk 18 above)
Comment (sl7v3) [inline app/views/friendships/_friends_list.html.erb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/c3f492188f15208b33325da7cd211ef506638973

File: app/views/friendships/_friends_list.html.erb (hunk 19 above)
Comment (sl7v3) [inline app/views/friendships/_friends_list.html.erb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/2a44cb24cd247c672ba64734360dc5314157ac98

File: app/views/friendships/edit.html.erb (hunk 20 above)
Comment (sl7v3) [inline app/views/friendships/edit.html.erb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/2a44cb24cd247c672ba64734360dc5314157ac98

File: app/views/friendships/index.html.erb (hunk 21 above)
Comment (sl7v3) [inline app/views/friendships/index.html.erb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/9205524d2dd94928dc799642ed1fcc643bd5c19d

File: test/fixtures/friendships.yml (hunk 22 above)
Comment (sl7v3) [inline test/fixtures/friendships.yml]: Believe it or not, this was auto-generated.

File: app/controllers/friendships_controller.rb (hunk 7 above)
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: I actually think user should be an instance variable.  The other changes are fine though.  If you make user a local variable then there is no benefit of doing `user ||= current_user` because `user` will have to redefined everytime a new action is carried out.  In reality, the same user will be  logged in regardless of the controller action executing and that variable is used in almost every one, so you would still have to do a db call everytime you want to get that data.   `@user ||= current_user` is the same as saying 
if @user == nil
  @user = current_user
We only want to do that db operation if we have not already defined the user.
Let me know if this doesn't make sense.

File: app/controllers/friendships_controller.rb (hunk 8 above)
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: Constants are specified with all caps.  We can ensure they are a constant by adding the `.freeze` method at the end of it.  `FRIENDS_LIST_LENGTH = 5.freeze`.  Also, you definitely don't want to make it a global variable since you will only be using it inside this class.  Just remove the $ sign and you should still have access to it anywhere in the controller since you declared it outside of a function.

File: db/migrate/20160726034132_add_last_logged_in_to_users.rb (hunk 10 above)
Comment (danfstucky) [inline db/migrate/20160726034132_add_last_logged_in_to_users.rb]: Alright, that makes sense.

File: app/models/friendship.rb (hunk 11 above)
Comment (danfstucky) [inline app/models/friendship.rb]: Oh, I learned something then.  I didn't realize `cattr_accessor` was a thing.  I assumed you meant to write `attr_accessor`.

File: app/models/friendship.rb (hunk 14 above)
Comment (danfstucky) [inline app/models/friendship.rb]: +1

File: app/models/user.rb (hunk 15 above)
Comment (danfstucky) [inline app/models/user.rb]: +1

File: app/views/friendships/_friends_list.html.erb (hunk 16 above)
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: +1

File: app/views/friendships/_friends_list.html.erb (hunk 17 above)
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: +1

File: app/views/friendships/_friends_list.html.erb (hunk 18 above)
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: +1

File: app/views/friendships/_friends_list.html.erb (hunk 19 above)
Comment (danfstucky) [inline app/views/friendships/_friends_list.html.erb]: +1

File: app/views/friendships/edit.html.erb (hunk 20 above)
Comment (danfstucky) [inline app/views/friendships/edit.html.erb]: +1

File: app/views/friendships/index.html.erb (hunk 21 above)
Comment (danfstucky) [inline app/views/friendships/index.html.erb]: +1

File: test/fixtures/friendships.yml (hunk 22 above)
Comment (danfstucky) [inline test/fixtures/friendships.yml]: That's strange.  I guess it doesn't matter. +1

File: app/models/friendship.rb (hunk 13 above)
Comment (danfstucky) [inline app/models/friendship.rb]: +1

File: app/models/friendship.rb (hunk 12 above)
Comment (danfstucky) [inline app/models/friendship.rb]: +1

File: app/controllers/search_controller.rb (hunk 9 above)
Comment (danfstucky) [inline app/controllers/search_controller.rb]: +q

File: app/controllers/friendships_controller.rb (hunk 6 above)
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: Not sure whats going on with indentation in github.  +1

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (danfstucky) [inline app/controllers/application_controller.rb]: +1

File: app/assets/stylesheets/friendships.scss (hunk 4 above)
Comment (danfstucky) [inline app/assets/stylesheets/friendships.scss]: +1

File: app/assets/stylesheets/friendships.scss (hunk 3 above)
Comment (danfstucky) [inline app/assets/stylesheets/friendships.scss]: +1

File: app/assets/stylesheets/application.css.scss (hunk 2 above)
Comment (danfstucky) [inline app/assets/stylesheets/application.css.scss]: +1

File: app/controllers/friendships_controller.rb (hunk 8 above)
Comment (sl7v3) [inline app/controllers/friendships_controller.rb]: Fixed here: https://github.com/danfstucky/LeagueSpec/commit/7acf7533ea4f2b2a6c5cc85e0edc6ab6f4c1f997

File: app/controllers/friendships_controller.rb (hunk 7 above)
Comment (sl7v3) [inline app/controllers/friendships_controller.rb]: You make a good point about the utility of the user variable versus @user variable and I agree. However, I was looking up the purpose of using instance variables after your first comment and the information I found summarized to: "Only use instance variables for variables that will be sent to the view." In the actions where I did not make user an instance variable it was because they were not being sent to the view.
However, since your explanation favors utility/less db access(which I am in support of) over convention, I went ahead and made the changes here:  https://github.com/danfstucky/LeagueSpec/commit/ac9e83afc4188b3a52cd57f4175c8f06c50e7e68

Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: Yes, in practice we should actually be storing that user variable completely somewhere else since it is used by multiple controllers.  I think this is fine for now.  It is generally alright to use instance variables for reasons other than passing info to views as long as it is a good reason.
+1

File: app/controllers/friendships_controller.rb (hunk 8 above)
Comment (danfstucky) [inline app/controllers/friendships_controller.rb]: +1

**Comment by danfstucky** [comment]
+1 Ready to merge.
You will definitely have some merge conflicts here so make sure you resolve them and test your changes before pushing to master.

**Comment by sl7v3** [comment]
Merged here: https://github.com/danfstucky/LeagueSpec/commit/56da23ee755029c8f5353c003fa96901c99c8f80
